primitives/orderbook/src/types.rs

```diff
@@ -0,0 +1,595 @@
+						Some(Ordering::Less)
+					},
+				// A.price > B.price => [A, B]
+				Ordering::Greater => Some(Ordering::Greater),
+			}
+		} else {
+			// Sell side
+			match self.price.cmp(&other.price) {
+				// A.price < B.price => [A, B] (in sell side, the first prices should be the lowest)
+				Ordering::Less => Some(Ordering::Greater),
+				// A.price == B.price => Order based on timestamp - lowest timestamp first
+				Ordering::Equal => {
+					// If price is equal, we follow the FIFO priority
+					if self.timestamp < other.timestamp {
+						Some(Ordering::Greater)
+					} else {
+						Some(Ordering::Less)
+					}
+				},
+				// A.price > B.price => [B, A]
+				Ordering::Greater => Some(Ordering::Less),
+			}
+		}
+	}
+}
+
+impl Ord for Order {
+	fn cmp(&self, other: &Self) -> Ordering {
+		assert_eq!(self.side, other.side, "Comparison cannot work for opposite order sides");
+		if self.side == OrderSide::Bid {
+			// Buy side
+			match self.price.cmp(&other.price) {
+				// A.price < B.price => [B, A] (in buy side, the first prices should be the highest)
+				Ordering::Less => Ordering::Less,
+				// A.price == B.price => Order based on timestamp
+				Ordering::Equal =>
+					if self.timestamp < other.timestamp {
+						Ordering::Greater
+					} else {
+						Ordering::Less
+					},
+				// A.price > B.price => [A, B]
+				Ordering::Greater => Ordering::Greater,
+			}
+		} else {
+			// Sell side
+			match self.price.cmp(&other.price) {
+				// A.price < B.price => [A, B] (in sell side, the first prices should be the lowest)
+				Ordering::Less => Ordering::Greater,
+				// A.price == B.price => Order based on timestamp
+				Ordering::Equal => {
+					// If price is equal, we follow the FIFO priority
+					if self.timestamp < other.timestamp {
+						Ordering::Greater
+					} else {
+						Ordering::Less
+					}
+				},
+				// A.price > B.price => [B, A]
+				Ordering::Greater => Ordering::Less,
+			}
+		}
+	}
+}
+
+impl Order {
+	/// Computes the new avg_price and adds qty to filled_qty
+	/// if returned is false then underflow occurred during division
+	pub fn update_avg_price_and_filled_qty(&mut self, price: Decimal, amount: Decimal) -> bool {
+		let mut temp = self.avg_filled_price.saturating_mul(self.filled_quantity);
+		temp = temp.saturating_add(amount.saturating_mul(price));
+		self.filled_quantity = self.filled_quantity.saturating_add(amount);
+		println!("self.filled_quantity: {:?}\ntemp: {:?}", self.filled_quantity, temp);
+		match temp.checked_div(self.filled_quantity) {
+			Some(quotient) => {
+				println!("Quotient: {:?}", quotient);
+				self.avg_filled_price = quotient;
+				true
+			},
+			None => false,
+		}
+	}
+
+	pub fn available_volume(&self, other_price: Option<Decimal>) -> Decimal {
+		//this if for market bid order
+		if self.qty.is_zero() {
+			println!(
+				"quote_order_qty: {:?}, avg_filled_price: {:?}, filled_quantity: {:?}",
+				self.quote_order_qty, self.avg_filled_price, self.filled_quantity
+			);
```

Comment:
Do we need this print? why not used logger?

Reply:
Same as above